browser/farbling/brave_speech_synthesis_farbling_browsertest.cc

```diff
@@ -0,0 +1,147 @@
+    browser_content_client_.reset(new BraveContentBrowserClient());
+    content::SetBrowserClientForTesting(browser_content_client_.get());
+
+    host_resolver()->AddRule("*", "127.0.0.1");
+    content::SetupCrossSiteRedirector(embedded_test_server());
+
+    brave::RegisterPathProvider();
+    base::FilePath test_data_dir;
+    base::PathService::Get(brave::DIR_TEST_DATA, &test_data_dir);
+    test_data_dir = test_data_dir.AppendASCII(kEmbeddedTestServerDirectory);
+    embedded_test_server()->ServeFilesFromDirectory(test_data_dir);
+
+    ASSERT_TRUE(embedded_test_server()->Start());
+  }
+
+  void TearDown() override {
+    browser_content_client_.reset();
+    content_client_.reset();
+  }
+
+  HostContentSettingsMap* content_settings() {
+    return HostContentSettingsMapFactory::GetForProfile(browser()->profile());
+  }
+
+  void AllowFingerprinting(std::string domain) {
+    brave_shields::SetFingerprintingControlType(
+        content_settings(), ControlType::ALLOW,
+        embedded_test_server()->GetURL(domain, "/"));
+  }
+
+  void BlockFingerprinting(std::string domain) {
+    brave_shields::SetFingerprintingControlType(
+        content_settings(), ControlType::BLOCK,
+        embedded_test_server()->GetURL(domain, "/"));
+  }
+
+  void SetFingerprintingDefault(std::string domain) {
+    brave_shields::SetFingerprintingControlType(
+        content_settings(), ControlType::DEFAULT,
+        embedded_test_server()->GetURL(domain, "/"));
+  }
+
+  content::WebContents* web_contents() {
+    return browser()->tab_strip_model()->GetActiveWebContents();
+  }
+
+  bool NavigateToURLUntilLoadStop(const GURL& url) {
+    EXPECT_TRUE(ui_test_utils::NavigateToURL(browser(), url));
+    return WaitForLoadStop(web_contents());
+  }
+
+ private:
+  std::unique_ptr<ChromeContentClient> content_client_;
+  std::unique_ptr<BraveContentBrowserClient> browser_content_client_;
+};
+
+// Tests results of farbling voices list
+IN_PROC_BROWSER_TEST_F(BraveSpeechSynthesisFarblingBrowserTest, FarbleVoices) {
+  std::string domain_b = "b.com";
+  std::string domain_z = "z.com";
+  GURL url_b =
+      embedded_test_server()->GetURL(domain_b, "/voices-farbling.html");
+  GURL url_z =
+      embedded_test_server()->GetURL(domain_z, "/voices-farbling.html");
+  // Farbling level: off
+  // The voices list should be the real voices list.
+  AllowFingerprinting(domain_b);
+  NavigateToURLUntilLoadStop(url_b);
+  std::string off_voices_b =
+      EvalJs(web_contents(), kTitleScript).ExtractString();
+  ASSERT_NE("failed", off_voices_b);
+  AllowFingerprinting(domain_z);
+  NavigateToURLUntilLoadStop(url_z);
+  std::string off_voices_z =
+      EvalJs(web_contents(), kTitleScript).ExtractString();
+  ASSERT_NE("failed", off_voices_z);
+  // The voices list should be the same on every domain if farbling is off.
+  EXPECT_EQ(off_voices_b, off_voices_z);
+
+  // On platforms without any voices, the rest of this test is invalid.
+  if (off_voices_b == "")
```

Comment:
this could hide the real issues. Can we do this check with platform-specific `ifdef`s?

Reply:
Unfortunately the API itself is so flaky (requires setting a timeout to decide that there are no voices) that the test sometimes fails on Windows CI where there are definitely voices. This was the compromise solution.

Reply:
Gotcha, but maybe it's the `title`-observing approach flaky in its nature? I was wondering why you decided to go with it instead of directly calling async JS via something like `content::EvalJS`.